[SE-NNNN] Allow Property Wrappers on Let Declarations (#1910)
swiftlang/swift-evolution: proposals/0383-allow-let-property-wrapper.md

# Allow Property Wrappers on Let Declarations

* Proposal: [SE-038N](NNNN-filename.md)
* Authors: [Amritpan Kaur](https://github.com/amritpan), [Pavel Yaskevich](https://github.com/xedin)
* Review Manager: TBD
* Status: **Awaiting review**
* Implementation: [apple/allow-let-property-wrapper](https://github.com/apple/swift/pull/62342) 

## Introduction

> SE-0258 Property Wrappers models both mutable and immutable variants of a property wrapper but does not allow wrapped properties explicitly marked as a `let`. This proposal builds on the original vision by expanding property wrapper usage to include let declared properties with a wrapper attribute.

Reviewer: ```suggestion
[SE-0258 Property Wrappers](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md) models both mutable and immutable variants of a property wrapper but does not allow wrapped properties explicitly marked as a `let`. This proposal builds on the original vision by expanding property wrapper usage to include let declared properties with a wrapper attribute.
```

## Motivation

> Allowing property wrappers to be applied to `let` properties improves Swift language consistency and code safety.

Reviewer: "Safety" in Swift Evolution lingo usually refers to memory safety, and we've discouraged looser usages. Do you mean here perhaps _expressivity_?

Author: I replaced "safety" with "expressivity" here. Is there also a difference between "expressivity" and "expressiveness" for Swift Evolution lingo?

Today, a property wrapper can be applied to a property of any type but the rules for declaring these wrapped properties are varied and singular to property wrappers. For example, a struct instance without the property wrapper attribute can be declared with either a `var` or a `let` property. However, mark the struct as a property wrapper type and the compiler no longer allows it to be written as a `let` wrapped property:

> ```

Reviewer: ```suggestion
```swift
```

@propertyWrapper
struct Wrapper {
  var wrappedValue: Int { 0 }
  init(wrappedValue: Int) {}
}

struct S {
  @Wrapper let value: Int // Error: Property wrapper can only be applied to a ‘var’
}
```
Permitting wrapped properties to mimic the rules for other type instances that can be written with either a `var` or a `let` will simplify the Swift language.

> Additionally, `let` wrapped properties add code safety where a user wants to expressly remove access to a property’s mutators after initializing the property once or simply does not need a mutable property wrapper. This could be useful for property wrappers that do not change or are reference types.

Reviewer: See comment above regarding the use of "safety."

Reviewer: Additionally, I would suggest consider incorporating the discussion in the pitch thread prompted by Gustaf's question: "What kinds of interesting behavior would this actually allow that we can't have now?"

The answers in the thread, beyond the point you mention above, are so far:
* Reference semantics if property wrapper implementation supports it (Pavel).
* `nonisolated` wrapped properties (Holly).

Although you mention reference semantics below, I would suggest it's worth calling it out here at the top. You may want to consider if there are other "interesting behaviors" that would fit the list.

## Proposed solution

We propose to allow the application of property wrappers to let declared properties, which will permit the wrapper type to be initialized only once without affecting the implementation of the backing storage.

> For example, [TSPL](https://docs.swift.org/swift-book/LanguageGuide/Properties.html#ID617) defines a `SmallNumber` property wrapper and applies it to `UnitRectangle` properties:

Reviewer: ```suggestion
For example, [_The Swift Programming Language_](https://docs.swift.org/swift-book/LanguageGuide/Properties.html#ID617) defines a `SmallNumber` property wrapper and applies it to `UnitRectangle` properties:
```

> ```

Reviewer: ```suggestion
```swift
```

@propertyWrapper
struct SmallNumber {
  private var maximum: Int
  private var number: Int

  var wrappedValue: Int {
    get { return number }
    set { number = min(newValue, maximum) }
  }

  init(wrappedValue: Int) {
    maximum = 12
    number = min(wrappedValue, maximum)
  }
}

struct UnitRectangle {
  @SmallNumber var height: Int = 1
  @SmallNumber var width: Int = 1
}
```
Initial values for `height` and `width` are set using the `init(wrappedValue:)` initializer of the `SmallNumber` property wrapper type. To ensure that `height` and `width` are not changed again, we could add logic to the `wrappedValue` setter to check if the property was already initialized and prevent re-assignment. However, this is an inconvenient solution.

Instead, we could declare these properties with a `let`, synthesize a local `let` constant for the backing storage property (prefixed with an underscore), and only allow the property wrapper to be initialized once, passing the assigned value to `init(wrappedValue:)`. Now, rewriting `UnitRectangle`’s properties as `let` constants will translate to:

> ```

Reviewer: ```suggestion
```swift
```

private let _height: SmallNumber = SmallNumber(wrappedValue: 1)
var height: Int {
  get { return _height.wrappedValue }
}

private let _weight: SmallNumber = SmallNumber(wrappedValue: 1)
var weight: Int {
  get { return _weight.wrappedValue }
}
```
and results in code that is easy to write and understand:

> ```

Reviewer: ```suggestion
```swift
```

struct UnitRectangle {
  @SmallNumber let height: Int = 1
  @SmallNumber let width: Int = 1
}
```

Property wrappers with `let` declarations will be allowed both as members and local declarations, as envisioned by SE-0258 for `var` declared property wrappers. All other property wrapper traits also remain unchanged from SE-0258.

## Detailed design

> Here are three examples of how a `let` wrapped property can make current iterations more effortless.

Reviewer: The examples are part "Motivation" and part "Proposed solution," but I don't think they fit with what's expected in this section.

Consider looking at what SE-0258 puts into this section for some ideas; it doesn't have to be so long of course. My suggestion would be that it bears formally spelling out that, just like before, property wrappers still work by transforming the source code, then showing worked examples of transforms using the surface language. Start with the basics that you demonstrate in the "Proposed solution" section, but work in the questions you've encountered on the pitch thread concerning how the transform will work in the case of `projectedValue`, `init()`, etc. Such questions dominated the existing discussion and it would be helpful that any document presented for review doesn't leave gaps in that understanding, as the review period will get more feedback evaluating the proposed design if it's not spent mostly developing readers' understanding of what that is.

In general, consider incorporating the questions on the pitch thread that are about "_what would happen if..._" in terms of the observed behavior if one tries to set some value directly or indirectly, or "_is it possible to..._" (for instance, the multiphase initialization example) as part of the "Proposed solution" section, and questions about _the mechanics of how_ some code is transformed as part of the "Detailed design."

> The `Clamping` property wrapper from [SE-0258 Property Wrappers](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md#clamping-a-value-within-bounds) can be rewritten to use `let` as its storage and implementation do not change, no matter its application.

Reviewer: What does the second half of this sentence mean? I feel like some words might be missing. How could the _implementation_ of the `Clamping` property wrapper change depending on its application?

> ```

Reviewer: ```suggestion
```swift
```

struct Color {
  @Clamping(min: 0, max: 255) let purple: Int = 127
}
```

A `let` wrapped property could be useful for reference types like a property wrapper class. Typically property wrappers are written for value types but occasionally a protocol like `NSObject` may require the use of a class. For example:

> ```

Reviewer: ```suggestion
```swift
```

@propertyWrapper
class WrapperClass<T> : NSObject {
  var wrappedValue: T

  init(wrappedValue: T) {
    self.wrappedValue = wrappedValue
  }
}

class C {
  @WrapperClass let value: Int

  init(v: Int) {
    value = v
  }
}
```
Here, `WrapperClass` can be made an immutable property wrapper class instance, preventing any future unintentional changes to the property wrapper class type in this context.

SwiftUI property wrappers may also benefit from a let declaration. For example, `@ScaledMetric` in its simplest usage can be written with a `let` instead:

> ```

Reviewer: ```suggestion
```swift
```

struct ContentView: View {
  @ScaledMetric let imageSize = 10

  var body: some View {
    Image(systemName: "heart.fill")
    .resizable()
    .frame(width: imageSize, height: imageSize)
  }
}
```

Similarly, other SwiftUI property wrappers could be `let` declared when they do not require more than a single initialization.

### Property wrappers with `nonmutating set`

Property wrappers that have a `wrappedValue` property with a `nonmutating set` (e.g., SwiftUI's [`@State`](https://developer.apple.com/documentation/swiftui/state/wrappedvalue) and [`@Binding`](https://developer.apple.com/documentation/swiftui/binding/wrappedvalue)) will preserve the reference semantics of the `wrappedValue` implementation even when marked as a `let` declaration. For example:  

> ```

Reviewer: ```suggestion
```swift
```

@State let weekday: String = "Monday"
```

> Here, `weekday` is an immutable instance of the `@State` property wrapper, but its `wrappedValue` storage will retain its mutability and reference type traits. This will translate to:

Reviewer: Neither `State` nor `String` is a reference type. How could this storage have "reference type traits"? Do you mean to say that the property wrapper will retain its reference _semantics_?

> ```

Reviewer: ```suggestion
```swift
```

private let _weekday: State<String> = State<String>(wrappedValue: "Monday")
var weekday : String {
  get { 
    return _weekday.wrappedValue 
  }
  nonmutating set {
    self._weekday.wrappedValue = value
  }
  nonmutating _modify { 
    yield () 
  }
}
var $weekday: Binding<String> {
  get { return _weekday.projectedValue }
}
```

Marking `weekday` as a let declared property will not remove access to its `wrappedValue`'s nonmutating set and the `wrappedValue` can be assigned via the backing property, `_weekday`:

> ```

Reviewer: ```suggestion
```swift
```

_weekday.wrappedValue = "Tuesday"
```

However, this does not affect the immutability of `weekday`, which can only be assigned to once like any ordinary let wrapped property. Any attempt to reassign `weekday` will result in an error:

> ```

Reviewer: ```suggestion
```swift
```

weekday = "Wednesday" // Error: Cannot assign to value: 'weekday' is a 'let' constant
```

## Source compatibility

This is an additive feature that does not impact source compatibility.

> ## Effect on ABI stability/API resilience

Reviewer: ```suggestion
## Effect on ABI stability
```

> This is an additive change that has no direct impact that compromises ABI stability.

Reviewer: In other reviews (maybe not this one), users have asked if new features can be backdeployed on older runtimes. It might be sufficiently obvious because property wrappers are a syntactic transformation, but for consistency it would be nice to comment here that no runtime support is needed for backdeployment.

## Effect on API resilience

> This is an additive change that has no impact on API resilience.

Reviewer: Are there "Alternatives considered" or "Future directions" applicable here? Or does this proposal really complete the picture in terms of what property wrappers can be expected to support for the foreseeable future?

Author: > In reviewing review thread comments, did you ever get around to looking into @ktoso's question about `@TaskLocal`?

Yes, I added [this test here](https://github.com/apple/swift/pull/62342/commits/a983c0740eef3beae00ea1c976a655f9fc2c0c68#diff-6808a8c1547f7f88cc7f05b0f122718c7388a69eed21d9a70b9873ca847f920b) to explicitly check for let wrapped TaskLocal.

You've given me quite a bit to rethink here - thank you! I'll work through this in the next day or so.

After some additional discussion about `nonmutating set`, I think I need to move this part of the proposal to Alternatives Considered section instead. I'll rework that this week as well. Requesting a few more days to work on this before we move forward.

Reviewer: Sure—thanks for being so responsive. Ping me when you're ready!

Similarly, other SwiftUI property wrappers could be `let` declared when they do not require more than a single initialization.

### Property wrappers with `nonmutating set`

Property wrappers that have a `wrappedValue` property with a `nonmutating set` (e.g., SwiftUI's [`@State`](https://developer.apple.com/documentation/swiftui/state/wrappedvalue) and [`@Binding`](https://developer.apple.com/documentation/swiftui/binding/wrappedvalue)) will preserve the reference semantics of the `wrappedValue` implementation even when marked as a `let` declaration. For example:  

```
@State let weekday: String = "Monday"
```
Here, `weekday` is an immutable instance of the `@State` property wrapper, but its `wrappedValue` storage will retain its mutability and reference type traits. This will translate to:
```
private let _weekday: State<String> = State<String>(wrappedValue: "Monday")
var weekday : String {
  get { 
    return _weekday.wrappedValue 
  }
  nonmutating set {
    self._weekday.wrappedValue = value
  }
  nonmutating _modify { 
    yield () 
  }
}
var $weekday: Binding<String> {
  get { return _weekday.projectedValue }
}
```

> Marking `weekday` as a let declared property will not remove access to its `wrappedValue`'s nonmutating set and `wrappedValue` can be assigned via the backing property:

Reviewer: To clarify, _only_ via the backing property, and _not_ directly by assigning to `weekday`, correct? The text should make this clear.

Author: Yes. I will make this clearer.
